### DISCUSSION TITLE
Architecture ...

### DISCUSSION THREAD

**Discussion by pheninux** [charmbracelet/bubbletea#286] [Q&A] [answered]
Hi
I wanted to know what is the right architecture to follow if you have a large project that contains several views with access to remote data or local database. thank you in advance

**Comment by tonyhb** [comment, marked as answer]
Hey Pheninux.  I'm not sure if you've ever looked into Elm or React, but if so it _might_ be similar.  Typically, you want:

1. A single top-level state manager.  This would contain an enum which controls each view.
2. Individual state structs for each view, embedded in the top-level state manager.

The top-level state manager is responsible for rendering all content.  It delegates to each embedded state store to render components, etc.  You have to pass down `tea.Msg` in `Update()` to all nested components.

RE. DB usage, you _can_ either:

1. Control DB outside of Tea via goroutines, and pass in messages using https://pkg.go.dev/github.com/charmbracelet/bubbletea?utm_source=godoc#Program.Send.
2. Control DB usage via the top-level `Update` commands.

Note that either way, your view only re-renders when Update is called (eg. a tea.Cmd is called which returns a tea.Msg).

There's a little bit of detail on multi-component architectures on this post here — it's a walkthrough on how to nest components with Bubbletea: https://www.inngest.com/blog/interactive-clis-with-bubbletea.

**Reply by pheninux** [reply]
hello Tonyhb ,
Thank you very much for the time you took to answer me.
